File: problems/205_isomorphic-strings/main.go
package main

func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	m1 := make(map[byte]byte)
	m2 := make(map[byte]byte)
	for i := range s {
		if bs, ok1 := m1[s[i]]; ok1 {
			if bt, ok2 := m2[t[i]]; ok2 {
				if bs != t[i] || bt != s[i] {
					return false
				}
			} else {
				return false
			}
		} else {
			if _, ok := m2[t[i]]; ok {
				return false
			}

			m1[s[i]] = t[i]
			m2[t[i]] = s[i]
		}
	}
	return true
}

File: problems/205_isomorphic-strings/__main__.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        s_t, t_s = {}, {}
        for ss, tt in zip(s, t):
            if ss not in s_t and tt not in t_s:
                s_t[ss] = tt
                t_s[tt] = ss
            elif ss in s_t and tt in t_s:
                if tt != s_t[ss] or ss != t_s[tt]:
                    return False
            else:
                return False
        return True
